vermouth/ffinput.py

```diff
@@ -51,196 +55,275 @@
+                ended.append(section.pop(-2))  # [a, b, c, d] -> [a, b, d]
+                wildcard_section = tuple(section[:-1] + ['*'])
+            self.section = section
+
+        result = None
+        if prev_section:
+            result = self.finalize_section(prev_section, ended)
+
+        action = self.header_actions.get(tuple(self.section))
+        if action is not None:
+            action()
+
+        return result
+
+    def parse_section(self, line, lineno):
+        """
+        Parse `line` with line number `lineno` by looking up the section in
+        :attr:`METH_DICT` and calling that method.
+
+        On the contrary to the SectionLineParser, we can have wildcard sections.
+        Wildcard sections cannot have children sections.
+
+        Parameters
+        ----------
+        line: str
+        lineno: int
+
+        Returns
+        -------
+        object
+            The result returned by calling the registered method.
+        """
+        line = _substitute_macros(line, self.macros)
+        wildcard_section = tuple(self.section[:-1] + ['*'])
+        end_section = []
+        if self.section:
+            end_section = self.section[-1]
+        if tuple(self.section) in self.METH_DICT:
+            method, kwargs = self.METH_DICT[tuple(self.section)]
+        elif wildcard_section in self.METH_DICT and end_section not in self.METH_DICT:
+            method, kwargs = self.METH_DICT[wildcard_section]
+        else:
+            raise IOError("Can't parse line {} in section '{}' because the "
+                          "section is unknown".format(lineno, self.section))
+        try:
+            return method(self, line, lineno, **kwargs)
+        except Exception as error:
+            raise IOError("Problems parsing line {}. I think it should be a "
+                          "'{}' line, but I can't parse it as such."
+                          "".format(lineno, self.section)) from error
+
+    def finalize(self, lineno=0):
+        for block in self.blocks.values():
+            # Because of hos they are described in gromacs, proper and improper
+            # dihedral angles are all under the [ dihedrals ] section. However
+            # the way they are treated differently in the library, at least on how
+            # they generate edges. Here we move the all the impropers into their own
+            # [ impropers ] section.
+            propers = []
+            impropers = []
+            for dihedral in block.interactions.get('dihedrals', []):
+                if dihedral.parameters and dihedral.parameters[0] == '2':
+                    impropers.append(dihedral)
+                else:
+                    propers.append(dihedral)
+            block.interactions['dihedrals'] = propers
+            block.interactions['impropers'] = impropers
+            block.make_edges_from_interactions()
+        for link in self.links:
+            link.make_edges_from_interactions()
```

Comment:
Don't we need the dihedral trick for Links? Or Modifications? The better place to address this would be in the parse_dihedral method, which I blindly assume exists.

Reply:
@pckroon Should we carry this one over to the new design base don an itp-reader , because it interprets GROMACS format?

Reply:
Sounds like a good idea to me.